File: bench/spectral-norm/java/SpectralNorm.java
// spectral-norm — see ../c/spectral-norm.c for the algorithm.
public final class SpectralNorm {
    static double a(int i, int j) {
        int s = i + j;
        return 1.0 / (s * (s + 1) / 2 + i + 1);
    }
    static void mulAv(double[] v, double[] out) {
        for (int i = 0; i < v.length; i++) {
            double s = 0.0;
            for (int j = 0; j < v.length; j++) s += a(i, j) * v[j];
            out[i] = s;
        }
    }
    static void mulAtv(double[] v, double[] out) {
        for (int i = 0; i < v.length; i++) {
            double s = 0.0;
            for (int j = 0; j < v.length; j++) s += a(j, i) * v[j];
            out[i] = s;
        }
    }
    static void mulAtAv(double[] v, double[] out, double[] tmp) {
        mulAv(v, tmp);
        mulAtv(tmp, out);
    }
    public static void main(String[] args) {
        int n = args.length > 0 ? Integer.parseInt(args[0]) : 100;
        double[] u = new double[n], v = new double[n], tmp = new double[n];
        java.util.Arrays.fill(u, 1.0);
        for (int i = 0; i < 10; i++) { mulAtAv(u, v, tmp); mulAtAv(v, u, tmp); }
        double vBv = 0.0, vv = 0.0;
        for (int i = 0; i < n; i++) { vBv += u[i] * v[i]; vv += v[i] * v[i]; }
        System.out.printf("%.9f\n", Math.sqrt(vBv / vv));
    }
}

File: bench/spectral-norm/rust/spectral-norm.rs
// spectral-norm — see ../c/spectral-norm.c for the algorithm.
fn a(i: usize, j: usize) -> f64 {
    let s = i + j;
    1.0 / ((s * (s + 1) / 2 + i + 1) as f64)
}
fn mul_av(v: &[f64], out: &mut [f64]) {
    let n = v.len();
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n { s += a(i, j) * v[j]; }
        out[i] = s;
    }
}
fn mul_atv(v: &[f64], out: &mut [f64]) {
    let n = v.len();
    for i in 0..n {
        let mut s = 0.0;
        for j in 0..n { s += a(j, i) * v[j]; }
        out[i] = s;
    }
}
fn mul_atav(v: &[f64], out: &mut [f64], tmp: &mut [f64]) {
    mul_av(v, tmp);
    mul_atv(tmp, out);
}
fn main() {
    let n: usize = std::env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(100);
    let mut u = vec![1.0f64; n];
    let mut v = vec![0.0f64; n];
    let mut tmp = vec![0.0f64; n];
    for _ in 0..10 {
        mul_atav(&u, &mut v, &mut tmp);
        mul_atav(&v, &mut u, &mut tmp);
    }
    let (mut vbv, mut vv) = (0.0, 0.0);
    for i in 0..n { vbv += u[i] * v[i]; vv += v[i] * v[i]; }
    println!("{:.9}", (vbv / vv).sqrt());
}
